reasoner/planner/AnswerCountEstimator.java

```diff
@@ -0,0 +1,486 @@
+            Set<LocalEstimate> subsetCoveredBy = coverMap.keySet().stream().filter(variablesToConsider::contains)
+                    .map(coverMap::get).collect(Collectors.toSet());
+            return subsetCoveredBy.stream().map(estimate -> estimate.answerEstimate(variablesToConsider)).reduce(1L, (x, y) -> x * y);
+        }
+
+        public boolean mayInitialize() {
+            return initializationStatus == InitializationStatus.NOT_STARTED || initializationStatus == InitializationStatus.RESET;
+        }
+
+        private void reset() {
+            assert this.initializationStatus == InitializationStatus.COMPLETE;
+            this.initializationStatus = InitializationStatus.RESET;
+            this.inferrableEstimates = null;
+            this.fullAnswerCount = -1;
+            this.negatedsCost = -1;
+        }
+
+        private void registerTriggeredRules() {
+            resolvables().filter(this::isInferrable).map(Resolvable::asConcludable).forEachRemaining(concludable -> {
+                Map<Rule, Set<Unifier>> unifiers = logicMgr.applicableRules(concludable);
+                iterate(unifiers.keySet()).forEachRemaining(rule -> {
+                    answerCountEstimator.registerConjunctionAndInitialize(rule.condition().conjunction());
+                });
+            });
+        }
+
+        private void initialize() {
+            if (!this.mayInitialize()) return;
+            //TODO: How does traversal handle type Variables?
+
+            initializationStatus = InitializationStatus.IN_PROGRESS;
+            // Create a baseline estimate from type information
+            if (unaryEstimateCover == null) unaryEstimateCover = computeNonInferredUnaryEstimateCover();
+
+            if (retrievableEstimates == null) retrievableEstimates = deriveEstimatesFromRetrievables();
+
+            registerTriggeredRules();
+
+            if (inferrableEstimates == null) {
+                inferrableEstimates = deriveEstimatesFromInferrables();
+                unaryEstimates = computeUnaryEstimatesWithInference(unaryEstimateCover);
+                unaryEstimateCover = computeFinalUnaryEstimateCover(unaryEstimateCover); // recompute with inferred
+            }
+
+            if (this.negatedsCost == -1) this.negatedsCost = computeNegatedsCost();
+
+            List<LocalEstimate> allEstimates = resolvables().flatMap(resolvable -> iterate(estimatesFromResolvable(resolvable))).toList();
+            Map<Variable, LocalEstimate> fullCostCover = computeGreedyEstimateCoverForVariables(allVariables(), allEstimates);
+            this.fullAnswerCount = costOfEstimateCover(allVariables(), fullCostCover) + this.negatedsCost;
+
+            // Can we prune the multiVarEstimates stored?
+            // Not if we want to order-resolvables. Else, yes based on queryableVariables.
+            initializationStatus = InitializationStatus.COMPLETE;
+        }
```

Comment:
Initialisation recursively computes all the rules and prepares all the LocalEstimates needed to answer queries. 
All recursion is handled at initialisation. The queries are answered without needing to recurse.